src/flamenco/runtime/program/fd_bpf_loader_v3_program.c

```diff
@@ -562,975 +581,1022 @@ process_loader_upgradeable_instruction( fd_exec_instr_ctx_t * instr_ctx ) {
+      /* Update the ProgramData account and record the program bits */
+      /* https://github.com/anza-xyz/agave/blob/574bae8fefc0ed256b55340b9d87b7689bcdf222/programs/bpf_loader/src/lib.rs#L669-L674 */
+      fd_borrowed_account_t * programdata = NULL;
+      err = fd_instr_borrowed_account_modify_idx( instr_ctx, 1UL, programdata_len, &programdata );
+      if( FD_UNLIKELY( err!=FD_ACC_MGR_SUCCESS ) ) {
+        FD_LOG_WARNING(( "Borrowed account lookup for programdata account failed" ));
+        return err;
+      }
 
-    /* rent is accessed directly from the epoch bank and the clock from the
-       slot context. However, a check must be done to make sure that the
-       sysvars are correctly included in the set of transaction accounts. */
-    err = fd_check_sysvar_account( instr_ctx, 4UL, &fd_sysvar_rent_id );
-    if( FD_UNLIKELY( err ) ) {
-      return err;
-    }
-    err = fd_check_sysvar_account( instr_ctx, 5UL, &fd_sysvar_clock_id );
-    if( FD_UNLIKELY( err ) ) {
-      return err;
-    }
+      fd_bpf_upgradeable_loader_state_t programdata_loader_state = {
+        .discriminant = fd_bpf_upgradeable_loader_state_enum_program_data,
+        .inner.program_data = {
+          .slot                      = clock.slot,
+          .upgrade_authority_address = (fd_pubkey_t *)authority_key,
+        },
+      };
+      err = fd_bpf_loader_v3_program_write_state( instr_ctx, 1UL, programdata, &programdata_loader_state );
+      if( FD_UNLIKELY( err!=FD_BINCODE_SUCCESS ) ) {
+        FD_LOG_WARNING(( "Bpf state write for programdata account failed" ));
+        return err;
+      }
+      FD_LOG_WARNING(("PERHAPS HERE BAD SIR?"));
 
-    if( FD_UNLIKELY( instr_ctx->instr->acct_cnt<7U ) ) {
-      FD_LOG_WARNING(( "Not enough account keys for instruction" ));
-      return FD_EXECUTOR_INSTR_ERR_NOT_ENOUGH_ACC_KEYS;
-    }
-    fd_pubkey_t const * authority_key = &txn_accs[ instr_acc_idxs[ 6UL ] ];
+      /* https://github.com/anza-xyz/agave/blob/574bae8fefc0ed256b55340b9d87b7689bcdf222/programs/bpf_loader/src/lib.rs#L675-L689 */
+      if( FD_UNLIKELY( PROGRAMDATA_METADATA_SIZE+buffer_data_len>programdata->meta->dlen ) ) {
+        uchar * sig = (uchar *)instr_ctx->txn_ctx->_txn_raw->raw + instr_ctx->txn_ctx->txn_descriptor->signature_off;
+        FD_LOG_WARNING(( "ProgramData account too small %32J %lu %lu", sig, buffer_data_len, programdata->meta->dlen ));
+        return FD_EXECUTOR_INSTR_ERR_ACC_DATA_TOO_SMALL;
+      }
+      if( FD_UNLIKELY( buffer_data_offset>buffer->meta->dlen ) ) {
+        FD_LOG_WARNING(( "Buffer account too small" ));
+        return FD_EXECUTOR_INSTR_ERR_ACC_DATA_TOO_SMALL;
+      }
 
-    /* https://github.com/anza-xyz/agave/blob/574bae8fefc0ed256b55340b9d87b7689bcdf222/programs/bpf_loader/src/lib.rs#L716-L745 */
-    /* Verify Program account */
+      /* Agave client calls get_data_mut on programdata data */
```

Comment:
Can we make the functions in fd_account.h line up with those in Agave's BorrowedAccount library, and use these instead? I think they should be 90% of the way there, we should fix this there instead.

For example:

- [BorrowedAccount::set_owner](https://github.com/anza-xyz/agave/blob/master/sdk/src/transaction_context.rs#L740-L766)
- [BorrowedAccount::set_lamports](https://github.com/anza-xyz/agave/blob/master/sdk/src/transaction_context.rs#L774-L796)
- [BorrowedAccount::checked_add_lamports](https://github.com/anza-xyz/agave/blob/master/sdk/src/transaction_context.rs#L798-L806)
- [BorrowedAccount::checked_sub_lamports](https://github.com/anza-xyz/agave/blob/master/sdk/src/transaction_context.rs#L808-L816)
etc...
etc...

We should use these functions everywhere in all the native programs.

Thank-you!!